File: Continuous-Subarray-Sum/CSS.py
class Solution(object):

    def checkSubarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        if len(nums) <= 1:
            return False
        elif sum(nums) == 0:
            return True
        elif k == 0:
            return False
        elif len(nums) >= 2 * k and k > 0:
            return True

        k = abs(k)
        result = {(0, 0): nums[0]}
        for count in range(2, len(nums) + 1):
            for i in range(0, len(nums) - count + 1):
                try:
                    tempRe = result[(i, i + count - 2)] + nums[i + count - 1]
                except:
                    tempRe = sum(nums[i:i + count])
                if tempRe % k == 0:
                    return True
                result[(i, i + count - 1)] = tempRe

        return False


a = Solution()
print(a.checkSubarraySum([23, 2, 4, 6, 7], 6))  # true
print(a.checkSubarraySum([23, 2, 6, 4, 7], -6))  # true
print(a.checkSubarraySum([0, 0], -1))  # true
print(a.checkSubarraySum([1, 1], 2))  # true
print(a.checkSubarraySum([1, 2, 3], 4))  # false
print(a.checkSubarraySum([1, 2, 3, 2, 10, 2], 14))  # true

File: Continuous-Subarray-Sum/CSS.rs
fn sum_vec(nums: &Vec<i32>) -> i32 {
    let mut result = 0;
    for i in nums {
        result += *i;
    }
    result
}

pub fn check_subarray_sum(nums: Vec<i32>, k: i32) -> bool {
    if nums.len() == 1 {
        return false;
    }
    let mut slice_copy = nums.as_slice();
    for count in 1..nums.len() {
        for start in 0..nums.len() - count {
            match k {
                0 => {
                    if slice_copy[start..start + count + 1].iter().sum::<i32>() == 0 {
                        return true;
                    }
                }
                _ => {
                    if slice_copy[start..start + count + 1].iter().sum::<i32>() % k.abs() == 0 {
                        return true;
                    }
                }
            }
        }
    }

    false
}

fn main() {
    dbg!(check_subarray_sum(vec![23, 2, 4, 6, 7], 6)); //t
    dbg!(check_subarray_sum(vec![23, 2, 6, 4, 7], -6)); //t
    dbg!(check_subarray_sum(vec![0, 0], -1)); //t
    dbg!(check_subarray_sum(vec![1, 2, 3], 4)); //nil
    dbg!(check_subarray_sum(vec![0, 0], 0)); //t
    dbg!(check_subarray_sum(vec![0], 0)); //nil
}
